drivers/adc/adc_mcp356x.c

```diff
@@ -0,0 +1,836 @@
+
+	/* Verify some things before we start programming the device registers */
+
+	if (cc->channel_id > 16 || cc->reference != ADC_REF_EXTERNAL0) {
+		return -EINVAL;
+	}
+	if (cc->input_positive >= 16 || cc->input_negative >= 16) {
+		return -EINVAL;
+	}
+	if (!cc->differential) {
+		LOG_WRN("The MCP356x only supports differential mode");
+		return -EINVAL;
+	}
+
+	uint8_t gain;
+
+	switch (cc->gain) {
+	case ADC_GAIN_1_3:
+		gain = REG_CONFIG2_GAIN_1_3;
+		break;
+	case ADC_GAIN_1:
+		gain = REG_CONFIG2_GAIN_1;
+		break;
+	case ADC_GAIN_2:
+		gain = REG_CONFIG2_GAIN_2;
+		break;
+	case ADC_GAIN_4:
+		gain = REG_CONFIG2_GAIN_4;
+		break;
+	case ADC_GAIN_8:
+		gain = REG_CONFIG2_GAIN_8;
+		break;
+	case ADC_GAIN_16:
+		gain = REG_CONFIG2_GAIN_16;
+		break;
+	case ADC_GAIN_32:
+		gain = REG_CONFIG2_GAIN_32;
+		break;
+	case ADC_GAIN_64:
+		gain = REG_CONFIG2_GAIN_64;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Inter-sample delay only configurable in scan mode.  One should use ΔΣ modulator
+	 * oversampling ratio to control this anyway.
+	 */
+	if (cc->channel_id == 16 &&
+	    cc->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		return -EINVAL;
+	}
+
+	if (cc->acquisition_time == ADC_ACQ_TIME_DEFAULT) {
+		data->delay = 0;
+	} else if (ADC_ACQ_TIME_UNIT(cc->acquisition_time) == ADC_ACQ_TIME_TICKS) {
+		unsigned int ticks = ADC_ACQ_TIME_VALUE(cc->acquisition_time);
+		/* Only powers of 2 between 8 and 512 inclusive */
+		if (ticks < 8 || ticks > 512 || (ticks & (ticks - 1))) {
+			return -EINVAL;
+		}
+		/* Only save upper 8 bits, need to shift back when used in adc read setup */
+		data->delay = REG_SCAN_DLY(ticks) >> 16;
+		LOG_DBG("delay of tics %u into bits %06x", ticks, REG_SCAN_DLY(ticks));
+	} else {
+		return -EINVAL;
+	}
+
+	/* Set Gain */
+	if (gain != data->gain) {
+		reg_write(dev, REG_CONFIG2, config->boost | gain |
+			  (data->az_mux ? REG_CONFIG2_AZ_MUX : 0) | REG_CONFIG2_RES);
+		data->gain = gain;
+	}
+
+	if (cc->channel_id == 16) {
+		if ((cc->input_positive < 8 && cc->input_positive >= config->channels) ||
+		    (cc->input_negative < 8 && cc->input_negative >= config->channels) ||
+		    cc->input_positive == 0xa || cc->input_negative == 0xa) {
+			return -EINVAL;
+		}
+
+		/* Set both muxes */
+		reg_write(dev, REG_MUX, cc->input_positive << 4 | cc->input_negative);
+	}
+	/* else, verify input settings are required values for channels 0-15? */
+
+	return 0;
+}
+
+static int mcp356x_read_async(const struct device *dev,
+			      const struct adc_sequence *sequence,
+			      struct k_poll_signal *async)
+{
+	const struct mcp356x_config *config = dev->config;
+	struct mcp356x_data *data = dev->data;
+	int err;
+
+	if (sequence->channels & BIT(16)) {
+		if (sequence->channels != BIT(16)) {
+			LOG_WRN("Channel 16 can't be sampled with additional channels");
+			return -EINVAL;
+		}
+		if (IS_ENABLED(CONFIG_ADC_MCP356X_USE_INTERNAL_TIMER) &&
+		    sequence->options && sequence->options->interval_us) {
```

Comment:
You should use essential boolean types for control statements (see MISRA-C rule 14.4, https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html).